In [1]:
import pandas as pd
import numpy as np

np.random.seed(123)

In [2]:
# Create list of values for names column.

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# Randomly generate arrays of scores for each student for each subject.
# Note that all the values need to have the same length here.

math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

In [3]:
# Construct the DataFrame using the above lists and arrays.

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades,
                   'classroom': np.random.choice(['A', 'B'], len(students))})

## Creating Dataframes

### From Lists and Dictionaries

There are several ways to create dataframes, we've already seen how we can create a dataframe from a dictionary:

In [4]:
pd.DataFrame({'A': [1,2,3], 'B': [4,5,6]})

A  B
0  1  4
1  2  5
2  3  6

The keys in the passed dictionary will be the column names, and the values are the data points that make up each column.

We can also create dataframes from a 2d data structure, either a numpy array or a list of lists:

In [5]:
pd.DataFrame([[1,2,3],[4,5,6]])

0  1  2
0  1  2  3
1  4  5  6

In [6]:
array = np.array([[1, 2, 3], [4, 5, 6]])

pd.DataFrame(array, columns=['a', 'b', 'c'])

a  b  c
0  1  2  3
1  4  5  6

Notice here that we had to specify the names of the columns ourselves.Notice here that we had to specify the names of the columns ourselves.

### From SQL

We can use the `read_sql` method to create a dataframe based on the results of a SQL query. To do this, we need to tell pandas how to connect to the database we are querying. The way we communicate this to pandas is with a specially formatted *connection string*.

In addition, whenever we want to connect to a database from our python code (other programming languages are similar), we will need a **driver**, a bit of software that handles the details of the database connection.

In order to connect to mysql, we'll install the `pymysql` driver packages:

`python -m pip install pymysql`

Once those are installed, we can create the connection string. In general, database connection urls will have this format:

`protocol://[user[:password]@]hostname/[database_name]`

Here's an example of what one would look like:

`mysql+pymysql://codeup:p@assw0rd@123.123.123.123/some_db`

Another thing we need to consider is that we don't want to publish our database credentials to github, however, we will need access to these values in our code in order to create the connection string defined above.

In order to accomplish this, we can define several variables in a file named `env.py` that contain the sensitive data, add `env.py` to our `.gitignore` file, and then import those values into another script.

**Be 100% sure to add `env.py` to this specific repository's `.gitignore` file, even and especially, if you have already added `env.py` to your global .gitignore file. This will protect the env file for people who clone this project (like collaborators)**

`from env import host, user, password`

`url = f'mysql+pymysql://{user}:{password}@{host}/employees'`

Now that the url is defined, we need to define our SQL query.

Once these two elements are defined, we can use them with the `read_sql` function to have pandas treat the results of a SQL query as a dataframe.

`query = 'SELECT * FROM employees LIMIT 5 OFFSET 50'`

`pd.read_sql(query, url)`

|  | emp_no | birth_date | first_name | last_name | gender | hire_date |
| --- | --- | --- | --- | --- | --- | --- |
| 0 | 10051 | 1953-07-28 | Hidefumi | Caine | M | 1992-10-15 |
| 1 | 10052 | 1961-02-26 | Heping | Nitsch | M | 1988-05-21 |
| 2 | 10053 | 1954-09-13 | Sanjiv | Zschoche | F | 1986-02-04 |
| 3 | 10054 | 1957-04-04 | Mayumi | Schueller | M | 1995-03-13 |
| 4 | 10055 | 1956-06-06 | Georgy | Dredge | M | 1992-04-27 |

It is common to have longer SQL queries that we want to read into python, and an example of how we might break a query into several lines is below:

In [7]:
# sql = '''
# SELECT
#     emp_no,
#     first_name,
#     last_name
# FROM employees
# WHERE gender = 'F'
# LIMIT 100
# '''

# employees = pd.read_sql(text(sql), engine.connect())
# employees.head()

**Passwords and Sensitive Information**

Don't add and commit files with passwords or other sensitive information in them to a git repository!

In [8]:
# query = '''
# SELECT
#     t.title as title,
#     d.dept_name as dept_name
# FROM titles t
# JOIN dept_emp USING (emp_no)
# JOIN departments d USING (dept_no)
# LIMIT 100
# '''

# title_dept = pd.read_sql(text(query), engine.connect())
# title_dept.head()

## Exercises I

Run `python -m pip install pymysql` from your terminal to install pymysql.

Create a notebook or python script named `advanced_dataframes` to do your work in for these exercises.

1. Run `python -m pip install pymysql` from your terminal to install the mysql client (any folder is fine)
2. cd into your exercises folder for this module and run `echo env.py >> .gitignore`
3. Create a function named `get_db_url`. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.
4. Use your function to obtain a connection to the `employees` database.
5. Once you have successfully run a query:
    
    a. Intentionally make a typo in the database url. What kind of error message do you see?
    
    b. Intentionally make an error in your SQL query. What does the error message look like?
    
6. Read the `employees` and `titles` tables into two separate DataFrames.
7. How many rows and columns do you have in each DataFrame? Is that what you expected?
8. Display the summary statistics for each DataFrame.
9. How many unique titles are in the `titles` DataFrame?
10. What is the oldest date in the `to_date` column?
11. What is the most recent date in the `to_date` column?

## Indexing and Subsetting

Like the pandas Series object, the pandas DataFrame object supports both position- and label-based indexing using the indexing operator `[]`.

I will demonstrate concrete examples of indexing using the indexing operator `[]` alone and with the `.loc` and `.iloc` attributes below.

### `[]`

I can pass a list of columns from a DataFrame to the indexing operator (aka bracket notation) to return a subset of my original DataFrame.

In [9]:
# Choose only two columns for my subset.

df[['name', 'classroom']]

name classroom
0     Sally         A
1      Jane         B
2     Suzie         A
3     Billy         B
4       Ada         A
5      John         B
6    Thomas         A
7     Marie         A
8    Albert         A
9   Richard         A
10    Isaac         B
11     Alan         A

In [10]:
# I can pass a boolean Series to the indexing operator as a selector.

bools = df.name.str.startswith('A')
bools

0     False
1     False
2     False
3     False
4      True
5     False
6     False
7     False
8      True
9     False
10    False
11     True
Name: name, dtype: bool

In [11]:
df[bools]

name  math  english  reading classroom
4      Ada    77       92       98         A
8   Albert    92       62       87         A
11    Alan    92       62       72         A

### `.loc`

We can use the `.loc` attribute to select specific rows AND columns by index label. The index label can be a number, but it can also be a string label. This method offers a lot of flexibility! **The .loc attribute's indexing is inclusive and uses an index label, not position.**

`df.loc[row_indexer, column_indexer]`

In [12]:
# Select all the rows and a subset of columns; notice the inclusive behavior of the indexing.

df.loc[:, 'math':'reading']

math  english  reading
0     62       85       80
1     88       79       67
2     94       74       95
3     98       96       88
4     77       92       98
5     79       76       93
6     82       64       81
7     93       63       90
8     92       62       87
9     69       80       94
10    92       99       93
11    92       62       72

In [13]:
# I can use a boolean Series as a selector with .loc, too, but I can choose rows and columns.

df.loc[bools, 'name': 'reading']

name  math  english  reading
4      Ada    77       92       98
8   Albert    92       62       87
11    Alan    92       62       72

### `.iloc`

We can use the `.iloc` attribute to select specific rows and colums by index position. `.iloc` does not accept a boolean Series as a selector like `.loc` does. **It takes in integers representing index position and is NOT inclusive.**

`df.iloc[row_indexer, column_indexer]`

We can select rows by integer position:

In [14]:
# Notice the exclusive behavior of the indexing.

df.iloc[:3]

name  math  english  reading classroom
0  Sally    62       85       80         A
1   Jane    88       79       67         B
2  Suzie    94       74       95         A

We can also specify which columns we want to select:

In [15]:
df.iloc[:3, 1:3]

math  english
0    62       85
1    88       79
2    94       74

Here we select the first 3 rows (everything up to but not including the index of 3), and the second and third columns (starting from the index of 1 up to but not including the index of 3).

## Aggregating

### `.agg`

The .agg method lets us specify a way to aggregate a series of numerical values. We pass an aggregate function or list of functions to the method that we want applied to a Series.

In [16]:
df.reading.agg('min')

67

While on the surface this seems pretty simple, .agg is capable of providing more detailed aggregations:

In [17]:
df[['english','reading','math']].agg(['mean','min','max'])

english  reading       math
mean  77.666667     86.5  84.833333
min   62.000000     67.0  62.000000
max   99.000000     98.0  98.000000

### `.groupby`

The `.groupby` method is used to create a grouped object, which we can then apply an aggregation on. For example, if we wanted to know the highest math grade from each classroom:

In [18]:
df.groupby('classroom').math.max()

classroom
A    94
B    98
Name: math, dtype: int64

We can use `.agg` here to, to see multiple aggregations:

In [19]:
df.groupby('classroom').math.agg(['min','mean','max'])

min    mean  max
classroom                  
A           62  82.625   94
B           79  89.250   98

We can group by multiple columns as well. To demonstrate, we'll create a boolean column named `passing_math`, then group by the combination of our new feature, `passing_math`, and the classroom and calculate the average reading grade and the number of individuals in each subgroup.

Let's break this problem down and code it step-by-step.

### **`np.where`**

First, we can create the new `passing_math` column using a handy NumPy function called `np.where`. It will allow us to base the new column values on whether the values in an existing column, `math`, meet a condition.

`np.where(condition, this_where_True, this_where_False)`

In [20]:
df['passing_math'] = np.where(df.math < 70, 'failing','passing')
df

name  math  english  reading classroom passing_math
0     Sally    62       85       80         A      failing
1      Jane    88       79       67         B      passing
2     Suzie    94       74       95         A      passing
3     Billy    98       96       88         B      passing
4       Ada    77       92       98         A      passing
5      John    79       76       93         B      passing
6    Thomas    82       64       81         A      passing
7     Marie    93       63       90         A      passing
8    Albert    92       62       87         A      passing
9   Richard    69       80       94         A      failing
10    Isaac    92       99       93         B      passing
11     Alan    92       62       72         A      passing

Next, we will group by the `passing_math` and `classroom` columns and use the `.agg` method to calculate the average reading grade and the number of students.

In [21]:
grade_groups = df.groupby(['passing_math','classroom']).reading.agg(['mean','count'])

In [22]:
grade_groups

mean  count
passing_math classroom                  
failing      A          87.000000      2
passing      A          87.166667      6
             B          85.250000      4

In [23]:
# I can even clean up my columns to make my calculations clearer.

grade_groups.columns = ['avg_reading_grade', 'count_of_students']
grade_groups

avg_reading_grade  count_of_students
passing_math classroom                                      
failing      A                  87.000000                  2
passing      A                  87.166667                  6
             B                  85.250000                  4

**Takeaways:**

We can interpret this output as there being 2 students failing math in classroom A with an average reading grade of 87, 6 students passing math in classroom A with an average reading grade of 87.16, and 4 students passing math in classroom B with an average reading grade of 85.25.

### **`.transform`**

The `.transform` method can be used to produce a series with the same length of the original dataframe where each value represents the aggregation from the subgroup resulting from the `.groupby`.

For example, if we wanted to know the average math grade for each classroom and add this data back to our original dataframe:

In [24]:
df.assign(avg_math_score_by_classroom=df.groupby('classroom').math.transform('mean'))

name  math  english  reading classroom passing_math  \
0     Sally    62       85       80         A      failing   
1      Jane    88       79       67         B      passing   
2     Suzie    94       74       95         A      passing   
3     Billy    98       96       88         B      passing   
4       Ada    77       92       98         A      passing   
5      John    79       76       93         B      passing   
6    Thomas    82       64       81         A      passing   
7     Marie    93       63       90         A      passing   
8    Albert    92       62       87         A      passing   
9   Richard    69       80       94         A      failing   
10    Isaac    92       99       93         B      passing   
11     Alan    92       62       72         A      passing   

    avg_math_score_by_classroom  
0                        82.625  
1                        89.250  
2                        82.625  
3                        89.250  
4                        82.625  
5                        89.250  
6                        82.625  
7                        82.625  
8                        82.625  
9                        82.625  
10                       89.250  
11                       82.625

### **`.describe`**

Check out what I can do when I combine a `.groupby` with a `.describe`!

In [25]:
df.groupby('classroom').reading.describe()

count    mean        std   min    25%   50%    75%   max
classroom                                                          
A            8.0  87.125   8.887190  72.0  80.75  88.5  94.25  98.0
B            4.0  85.250  12.392874  67.0  82.75  90.5  93.00  93.0

## Merging and Joining

Pandas provides several ways to combine dataframes together. We will look at two of them below:

### `pd.concat`

This function takes in a list or dictionary of Series or DataFrame objects and joins them along a particular axis, row-wise axis=0 or column-wise axis=1.

# For example, concat with a list of two DataFrames
pd.concat([df1, df2], axis=0)

- When your list contains at least one DataFrame, a DataFrame is returned.
- When concatenating only Series objects row-wise, axis=0, a Series is returned.
- When concatenating Series or DataFrames column-wise, axis=1, a DataFrame is returned.

# Default is set to row-wise concatenation using an outer join.
pd.concat(objs, axis=0, join='outer')

When concatenating dataframes vertically, we basically are just adding more rows to an existing dataframe. In this case, the dataframes we are putting together should have the same column names1.

In [26]:
df1 = pd.DataFrame({'a': [1, 2, 3]})
df2 = pd.DataFrame({'a': [4, 5, 6]})

df1

a
0  1
1  2
2  3

In [27]:
df2

a
0  4
1  5
2  6

In [28]:
pd.concat([df1,df2])

a
0  1
1  2
2  3
0  4
1  5
2  6

Note that the indices are preserved on the resulting dataframe; we could set the `ignore_index` parameter to `True` if we wanted these to be sequential.

In [29]:
concat_df1 = pd.concat([df1, df2], ignore_index=True)
concat_df1

a
0  1
1  2
2  3
3  4
4  5
5  6

In [30]:
concat_df2 = pd.DataFrame({'b': [1, 2, 3, 4, 5, 6]})
concat_df2

b
0  1
1  2
2  3
3  4
4  5
5  6

In [31]:
pd.concat([concat_df1, concat_df2], axis=1)

a  b
0  1  1
1  2  2
2  3  3
3  4  4
4  5  5
5  6  6

### `.merge`

This method is similar to a SQL join. Here's a [cool read](https://pandas.pydata.org/pandas-docs/stable/getting_started/comparison/comparison_with_sql.html#compare-with-sql-join) making a comparison between the two, if you're interested.

# df.merge default settings for commonly used parameters.

left_df.merge(right_df, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False,indicator=False)

How does changing the default argument of the `how` parameter change my resulting DataFrame?

`how` == Type of merge to be performed.

`how=left`: use only keys from left frame, similar to a SQL left outer join; preserve key order.

`how=right`: use only keys from right frame, similar to a SQL right outer join; preserve key order.

`how=outer`: use union of keys from both frames, similar to a SQL full outer join; sort keys lexicographically.

`how=inner`: use intersection of keys from both frames, similar to a SQL inner join; preserve the order of the left keys.

In [32]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [33]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

The `.merge` method will allow us to specify `left_on` and `right_on` to indicate the columns that are the keys used to merge the dataframes together.

- In addition, the `how` keyword argument is used to define what type of JOIN we want to do; as we saw above, `inner` is the default setting.
- For demonstration purposes, I'm also going to set the `indicator` parameter to `True`, which will create a column indicating whether the merge key appears in the `left_only`, `right_only` or `both` DataFrames.

In [34]:
# Perform an outer join specifying the left and right DataFrame keys.

users.merge(roles, left_on='role_id', right_on='id', how='outer', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

Notice that we have duplicate column names in the resulting dataframe. By default, pandas will add a suffix of `_x` to any columns in the left dataframe that are duplicated, and `_y` to any columns in the right dataframe that are duplicated. I can clean up my columns if I want to; one way would be to use method chaining, which it demonstrated below:

In [35]:
(users.merge(roles, 
            left_on='role_id', 
            right_on='id', 
            how='outer')
    .drop(columns='role_id')
    .rename(columns={'id_x': 'id', 
                     'name_x': 'employee',
                     'id_y': 'role_id',
                     'name_y': 'role'}
            )
)

id employee  role_id       role
0  1.0      bob      1.0      admin
1  2.0      joe      2.0     author
2  3.0    sally      3.0   reviewer
3  4.0     adam      3.0   reviewer
4  5.0     jane      NaN        NaN
5  6.0     mike      NaN        NaN
6  NaN      NaN      4.0  commenter

## Exercises II

1. Copy the `users` and `roles` DataFrames from the examples above.
2. What is the result of using a `right` join on the DataFrames?
3. What is the result of using an `outer` join on the DataFrames?
4. What happens if you drop the foreign keys from the DataFrames and try to merge them?
5. Load the `mpg` dataset from PyDataset.
6. Output and read the documentation for the `mpg` dataset.
7. How many rows and columns are in the dataset?
8. Check out your column names and perform any cleanup you may want on them.
9. Display the summary statistics for the dataset.
10. How many different manufacturers are there?
11. How many different models are there?
12. Create a column named `mileage_difference` like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.
13. Create a column named `average_mileage` like you did in the DataFrames exercises; this is the mean of the city and highway mileage.
14. Create a new column on the `mpg` dataset named `is_automatic` that holds boolean values denoting whether the car has an automatic transmission.
15. Using the `mpg` dataset, find out which which manufacturer has the best miles per gallon on average?
16. Do automatic or manual cars have better miles per gallon?